pkg/loki/flow_query.go

```diff
@@ -24,36 +24,44 @@ var filterRegexpValidation = regexp.MustCompile(`^[\w-_.,\"*:/]*$`)
 
 // FlowQueryBuilder stores a state to build a LogQL query
 type FlowQueryBuilder struct {
-	config       *Config
-	startTime    string
-	endTime      string
-	limit        string
-	labelFilters []labelFilter
-	lineFilters  []lineFilter
-	jsonFilters  [][]labelFilter
+	config           *Config
+	startTime        string
+	endTime          string
+	limit            string
+	labelFilters     []labelFilter
+	lineFilters      []lineFilter
+	extraLineFilters []string
+	jsonFilters      [][]labelFilter
 }
 
-func NewFlowQueryBuilder(cfg *Config, start, end, limit, reporter string) *FlowQueryBuilder {
+func NewFlowQueryBuilder(cfg *Config, start, end, limit, reporter, layer string) *FlowQueryBuilder {
 	// Always use app stream selector, which will apply whichever matching criteria (any or all)
 	labelFilters := []labelFilter{
 		stringLabelFilter(constants.AppLabel, constants.AppLabelValue),
 	}
+	extraLineFilters := []string{}
 	if reporter == constants.ReporterSource {
 		labelFilters = append(labelFilters, stringLabelFilter(fields.FlowDirection, "1"))
 	} else if reporter == constants.ReporterDestination {
 		labelFilters = append(labelFilters, stringLabelFilter(fields.FlowDirection, "0"))
 	}
+	if layer == constants.LayerInfrastructure {
```

Comment:
did yo try to do that filtering in the stream selector? Like always, to use indexes ..  you know it's my obsession 😈

Reply:
I guess we could but I have an issue here. Correct me if I'm wrong but labels doesn't manage `and` / `or` operators so we need to merge regex:

- Infra:
`SrcK8S_Namespace=~"^$|(kube-|openshift-).*|.*-ingress$"`

- App: 
`SrcK8S_Namespace=~"^$|^(?!kube-|openshift-).*$|.*-ingress$"`

These works fine in [regextester.com](https://www.regextester.com) but I get an error with the negative lookahead running app regex in Loki:
```
parse error : error parsing regexp: invalid or unsupported Perl syntax: `(?!`
```

Reply:
@jotak can we merge this as is ?

Reply:
OK but can we get some query perf measurements, to decide if it's something to improve asap or not?
cc  @memodi
Just to make sure there's no (or negligible) regressions on that point.

Reply:
sure I'll take a look :+1:

Reply:
What's "exec time" compared to the other time measurements?

I've also checked using our prom metrics and also see the increase when using apps or infra filtering, and indeed I see an important increase (in average time and also p50 / p95), between x2 and x4, when filtering on apps or infra, similar to what you've seen.

So, maybe we should at least set the "both layers" mode as a default, until we fix the performance drop, don't you think? So that it will be unnoticed unless a filter is applied. We can add a story to the global perf epic then.

Reply:
`exec time` is returned by loki in the `stats` section of each query. 

I'm totally fine to set `both` as default. I'll create the followup.

Thanks @jotak